### PR TITLE
feat: authenticator

### PR DESCRIPTION
**Pull request by brendt**

### CODE REVIEW

**Comment by brendt** [comment]
After this PR, we'll look into an authorization API (preferably with user roles built-in). All front-end stuff and user-management is left alone for now.

File: src/Tempest/Auth/.gitattributes [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# Exclude build/test files from the release
+.github/                  export-ignore
+tests/                    export-ignore
+.gitattributes            export-ignore
+.gitignore                export-ignore
+phpunit.xml               export-ignore
+README.md                 export-ignore
+
+# Configure diff output for .php and .phar files.
+*.php diff=php
```
Comment (Treggats) [inline src/Tempest/Auth/.gitattributes:10]:
```suggestion
*.php diff=php

```

File: src/Tempest/Auth/composer.json [hunk 2]
```diff
@@ -0,0 +1,12 @@
+{
+    "name": "tempest/auth",
+    "require": {
```
Comment (Treggats) [inline src/Tempest/Auth/composer.json:3]: You'd probable have a better description than this 😅 

```suggestion
    "name": "tempest/auth",
    "description": "Tempest authentication",
    "require": {
```

File: src/Tempest/Database/src/database.sqlite
Comment (Treggats) [inline src/Tempest/Database/src/database.sqlite:1]: I'm not a fan having this in the git tree, why did you include it?

File: src/Tempest/Framework/Testing/Http/TestResponseHelper.php [hunk 3]
```diff
@@ -177,9 +180,35 @@ public function assertHasNoValidationsErrors(): self
             sprintf(
                 "There should be no validation errors, but there were: %s",
                 implode(', ', array_keys($validationErrors)),
-            )
+            ),
```
Comment (Treggats) [inline src/Tempest/Framework/Testing/Http/TestResponseHelper.php:183]: did you add this purely because of the possible `message` argument?

**Review by Treggats — CHANGES_REQUESTED** [review]

**Comment by innocenzi** [comment]
I think this is great, there shouldn't need to be more built-in (besides authorization) 👍

However, I wouldn't let Tempest take care of the user model itself and the migration, I would rather have Tempest use an interface that has to be implemented in userland

File: src/Tempest/Database/src/database.sqlite
Comment (brendt) [reply src/Tempest/Database/src/database.sqlite:1]: IDK actually 🤔

File: src/Tempest/Framework/Testing/Http/TestResponseHelper.php (hunk 3 above)
Comment (brendt) [reply src/Tempest/Framework/Testing/Http/TestResponseHelper.php:183]: This was my autoformatter, I don't know why php cs fixer didn't pick up on it though

**Comment by brendt** [comment]
> However, I wouldn't let Tempest take care of the user model itself and the migration, I would rather have Tempest use an interface that has to be implemented in userland

Yeah, that's a fair point.

**Comment by brendt** [comment]
While I want the user model to be configurable, I do want to provide a default implementation — I might reconsider that though. 

Also, I'm ok with "the user" being tied to the database, I don't really care about other ways of storing user data for now.

**Comment by innocenzi** [comment]
Looking at the current code, it's unclear to me—is it possible to ignore the provided migration? Or would you recommend to add another one that would run after it and modify the table as we see fit?

That'd be one reason I was suggesting not to provide a default model/migration, but it seems low enough friction

**Comment by aidan-casey** [comment]
I'd love to see this use a repository to decouple retrieving the user details. If it were me personally, my criteria would be for this to be decoupled so that it could work entirely outside the framework. I understand that's a bit of an extreme view, though. 🙂

**Comment by brendt** [comment]
> Looking at the current code, it's unclear to me—is it possible to ignore the provided migration? Or would you recommend to add another one that would run after it and modify the table as we see fit?

Right now it's not possible, but this is something I already planned on figuring out before merging this PR :)

**Comment by brendt** [comment]
> I understand that's a bit of an extreme view, though. 🙂

Yes, not within my scope for now :)

**Comment by brendt** [comment]
> We have to think about the people storing passwords in the filesystem, @brendt!

I thought about it: they are free to roll their own auth layer separate from Tempest's default implementation 👍

File: src/Tempest/Auth/phpunit.xml [hunk 4]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="https://schema.phpunit.de/11.4/phpunit.xsd" bootstrap="vendor/autoload.php" executionOrder="depends,defects" beStrictAboutOutputDuringTests="true" displayDetailsOnPhpunitDeprecations="true" failOnPhpunitDeprecation="false" failOnRisky="true" failOnWarning="true">
+    <testsuites>
+        <testsuite name="Tempest Clock">
```
Comment (aidan-casey) [inline src/Tempest/Auth/phpunit.xml:4]: Update the package name. 😉

File: src/Tempest/Auth/composer.json [hunk 5]
```diff
@@ -0,0 +1,13 @@
+{
+    "name": "tempest/auth",
+    "description": "Tempest authentication",
```
Comment (aidan-casey) [inline src/Tempest/Auth/composer.json:3]: Update to something like:

> A flexible authentication package for the Tempest PHP framework, providing user authentication, session management, and token-based access control.

File: src/Tempest/Auth/src/User.php [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+use BackedEnum;
+use SensitiveParameter;
+use Tempest\Database\DatabaseModel;
+use Tempest\Database\IsDatabaseModel;
+use function Tempest\Support\arr;
+use UnitEnum;
+
+final class User implements DatabaseModel, CanAuthenticate, CanAuthorize
+{
+    use IsDatabaseModel;
```
Comment (aidan-casey) [inline src/Tempest/Auth/src/User.php:16]: :sob:

File: src/Tempest/Database/src/Id.php [hunk 7]
```diff
@@ -15,11 +15,18 @@
 
     public function __construct(string|int|self $id)
```
Comment (aidan-casey) [inline src/Tempest/Database/src/Id.php:16]: Is there a reason we don't separate these out?

File: src/Tempest/Database/src/Mappers/QueryToModelMapper.php [hunk 8]
```diff
@@ -51,18 +51,40 @@ private function parse(ClassReflector $class, DatabaseModel $model, array $row):
 
             $count = count($keyParts);
 
+            // TODO: clean up and document
```
Comment (aidan-casey) [inline src/Tempest/Database/src/Mappers/QueryToModelMapper.php:54]: Do et. 🙃

File: src/Tempest/Auth/src/CanAuthorize.php [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+
+declare(strict_types=1);
+
+namespace Tempest\Auth;
+
+use UnitEnum;
+
+interface CanAuthorize
+{
+    public function grantPermission(string|UnitEnum $permission): self;
```
Comment (aidan-casey) [inline src/Tempest/Auth/src/CanAuthorize.php:11]: Should the Authorizer really have this?

Comment (brendt) [reply src/Tempest/Auth/src/CanAuthorize.php:11]: Not sure yet, I'm in the sketching phase right now :)

File: src/Tempest/Database/src/Mappers/QueryToModelMapper.php (hunk 8 above)
Comment (brendt) [reply src/Tempest/Database/src/Mappers/QueryToModelMapper.php:54]: I will before merging!

File: src/Tempest/Database/src/Id.php (hunk 7 above)
Comment (brendt) [reply src/Tempest/Database/src/Id.php:16]: The idea is that all IDs are represented by a single value object, and that people don't have to worry about using UUIDs or normal ids. BUT that's in theory, I haven't really touched this code in ages. But outside the scope of this PR

File: src/Tempest/Auth/src/User.php (hunk 6 above)
Comment (brendt) [reply src/Tempest/Auth/src/User.php:16]: I might refactor it! Step by step

**Comment by brendt** [comment]
How to deal with this circular dependency: auth requires http because of session management, but http also requires auth because of the middleware. 

Since this auth layer is only concerned with http authentication, would it make sense to keep it all within the same component?

I know @aidan-casey will tell me otherwise :)

**Comment by innocenzi** [comment]
Is the circular dependency itself a technical issue? If not, I think keeping the `tempest/auth` package separate is a good idea

**Comment by brendt** [comment]
Apparently it is, composer can't deal with it (at least not in our CI setup). There's no issue in this branch because I haven't added the circular dependency yet, but we had the same issue once before between core and event-bus. 

Maybe there's a solution for it, I'll do some digging.

**Comment by aidan-casey** [comment]
@brendt - I haven't looked through the code yet, but probably I would have said use a driver/repository for storing the auth data. 🙈 🙃

**Comment by brendt** [comment]
I solved the issue, it was pretty straight forward: just need to hook into the Kernel::BOOTED event and add the middleware like so

File: src/Tempest/Database/src/Mappers/QueryToModelMapper.php (hunk 8 above)
Comment (brendt) [reply src/Tempest/Database/src/Mappers/QueryToModelMapper.php:54]: Ok I won't before merging 🫣

I don't really like this part, and I need to think some more about how to improve it. I made an issue for it though: #507

**Comment by brendt** [comment]
Ok so, this branch contains so much code that I want to merge it rather sooner than later. I'm still going to write an implementation for `revokePermission`, but after that I plan on merging it.

**Pull request merged**